src/manager/msNotificationManager.ts

```diff
@@ -27,47 +32,72 @@ export class MsNotificationManager extends MsManager {
-            ),
-        });
+        const message = Soup.Message.new(
+            'GET',
+            `${API_SERVER}/notifications?lastCheck=${previousCheck.toISOString()}`
+        );
+
         // send the HTTP request and wait for response
-        this.httpSession.queue_message(message, () => {
-            if (message.status_code != Soup.KnownStatusCode.OK) {
-                global.log(
-                    `error fetching notification ${message.status_code.toString()}`
-                );
-                return;
-            }
-
-            let notifications: NotificationResponseItem[] = [];
-            try {
-                notifications = JSON.parse(message.response_body.data);
-            } catch (e: unknown) {
-                global.log(`error unpack notification error ${e}`);
-                return;
-            }
-            const source = new MsNotificationSource();
-            notifications.forEach((notificationData) => {
-                Main.messageTray.add(source);
-                const notification = new MsNotification(
-                    source,
-                    notificationData.title,
-                    notificationData.content,
-                    notificationData.icon,
-                    notificationData.action
-                );
-
-                source.showNotification(notification);
+        // Before gnome 43 we use Soup 2.4 API
+        if (compareVersions(gnomeVersionNumber, parseVersion('43.0')) < 0) {
+            this.httpSession.queue_message(message, () => {
+                if (message.status_code != Soup.KnownStatusCode.OK) {
+                    global.log(
+                        `error fetching notification ${message.status_code.toString()}`
+                    );
+                    return;
+                }
+
+                let notifications: NotificationResponseItem[] = [];
+                try {
+                    notifications = JSON.parse(message.response_body.data);
+                } catch (e: unknown) {
+                    global.log(`error unpack notification error ${e}`);
+                    return;
+                }
+                this.showNotifications(notifications);
             });
-        });
+        } else {
+            this.httpSession.send_and_read_async(
+                message,
+                GLib.PRIORITY_DEFAULT,
+                new Gio.Cancellable(),
+                (session, result) => {
+                    if (session && message.status_code === Soup.Status.OK) {
+                        const bytes = session.send_and_read_finish(
+                            result
+                        ) as GLib.Bytes;
+                        const decoder = new TextDecoder('utf-8');
+                        const response = decoder.decode(
+                            bytes.get_data() as ArrayBuffer
+                        );
+                        this.showNotifications(
```

Comment:
'this' can have been destroyed when the callback is called. You can either cancel the callback or do a check inside it.

Reply:
Fair enough.